users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -0,0 +1,300 @@
+                    tf.expand_dims(tf.expand_dims(tf.less(i, num), axis=-1), axis=-1),
+                    _mask(
+                        x,
+                        batch_axis=batch_axis,
+                        axis=axis,
+                        pos=indices[:, i],
+                        max_amount=max_dims,
+                        sorted_indices=sorted_indices,
+                    ),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def transform(data, network, **config):
+    import tensorflow as tf
+
+    x = data.placeholder
+    step = network.global_train_step
+    current_epoch = tf.cast(step / config["steps_per_epoch"], tf.int32)
+
+    specaug_params = config["specaug_params"]
+
+    # Determine if we should use sorting
+    use_sorting = tf.logical_and(
+        config["enable_sorting"], tf.greater_equal(current_epoch, config["sorting_start_epoch"])
+    )
+
+    # Get filter indices (sorted or unsorted)
+    filter_layer = network.layers["features"].subnetwork.layers["conv_h_filter"].output.placeholder
+    num_filters = network.layers["features"].subnetwork.layers["conv_l"].output.shape[-1]
+
+    def get_sorted_indices():
+        sorted_filter_indices = sort_filters_by_center_freq(filter_layer)
+        sorted_indices = tf.stack(
+            [sorted_filter_indices * num_filters + filter_idx for filter_idx in range(num_filters)]
+        )
+        return tf.reshape(tf.transpose(sorted_indices), (-1,))
+
+    def get_unsorted_indices():
+        return tf.range(tf.shape(x)[data.feature_dim_axis])
+
+    sorted_indices = tf.cond(use_sorting, get_sorted_indices, get_unsorted_indices)
+
+    def get_masked():
+        x_masked = x
+        time_mask_max_size = tf.gather(specaug_params["time_mask_max_size"], current_epoch)
+        time_mask_max_num = tf.gather(specaug_params["time_mask_max_num"], current_epoch)
+        freq_mask_max_num = tf.gather(specaug_params["freq_mask_max_num"], current_epoch)
+        freq_mask_max_size = tf.gather(specaug_params["freq_mask_max_size"], current_epoch)
+        time_mask_max_proportion = tf.cast(
+            tf.math.floor(config["time_mask_max_proportion"] * tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        freq_mask_max_proportion = tf.cast(
+            tf.math.floor(config["freq_mask_max_proportion"] * tf.cast(tf.shape(x)[data.feature_dim_axis], tf.float32)),
+            tf.int32,
+        )
+
+        # check for the limits
+        actual_time_mask_max_num = tf.minimum(time_mask_max_num, time_mask_max_proportion // time_mask_max_size)
+        actual_freq_mask_max_num = tf.minimum(freq_mask_max_num, freq_mask_max_proportion // freq_mask_max_size)
+
+        x_masked = random_mask(
+            x_masked,
+            batch_axis=data.batch_dim_axis,
+            axis=data.time_dim_axis,
+            min_num=0,
+            max_num=actual_time_mask_max_num,
```

Comment:
This is different from the existing implementation which is
```
max_num=tf.maximum(
    tf.shape(x)[data.time_dim_axis] // int(1.0 / 0.7 * max_time),
    max_time_num,
) // (1 + increase_flag)
```
You don't have this `increase_flag` to reduce the number at the beginning of the training (need to check for stability). Also the first part of the maximum in the existing variant is missing. Did you think about what effect this had and how this changes in your case? This might be critical.

Reply:
Yes, I tried to remove as much hard-coded things as possible from the SpecAugment implementation. This gets replaced by the scheduling and `freq_mask_max_proportion` and  `time_mask_max_proportion` replace the maximum. This way I can set a upper limit for the masks and change the mask growths as well. I thought about making the randomness of the mask sizes and numbers also configurable but decided against it for simplicity.

Reply:
I see that you have similar options for configuration. I just feat that your current version results in very different behaviour compared to the existing version. This would introduce significant tuning effort for your version. Previously, `max_num` was the maximum of `max_time_num` and a term that grows with the sequence length. In your case, you take the minimum.

Reply:
I really want to avoid hidden scheduling in the mask sizes when I have an explicit scheduling now. This will need some new tuning, but tuning the masks for SCF was the goal of this implementation anyway.

Reply:
I see this point. But can you then please provide me with some simulations (like a plot max_num over epochs) on how this develops over the training with the old baseline implementation and your current baseline?

Reply:
![plot](https://github.com/user-attachments/assets/c71fb31d-d86a-4083-bc20-21e9fa6bf6bf)
Without scheduling it is missing the slow start for the first 2000 steps. 
The same with the current scheduling that I added but here the number of masks also gets bigger after 5 epochs. I could change the default scheduling to go from 2 to 5 after one epoch. That would be the closest I can get to the old behavior with the new Implementation

Reply:
Can you please compare original with your implementation without schedule for different sequence lengths? This is where things become interesting.

Reply:
![plot_time](https://github.com/user-attachments/assets/2a5519cf-8521-408d-9cac-128063393f80)
I see the Problem. The old implementation scaled with the sequence length while the new one will stay static (3) independent of the input length. 
- I could add a multiplicator to the number of masks based on the sequence length (fixed or changeable)
- I could make it a new parameter (minimal proportion) similar to the maximum proportion 
- I could make it a fixed minimum percentile like it is in the old implementation

Reply:
This is not at all about a minimum. The minimum number of masks is always 0 in the existing implementation. This is about the maximum only. And in the existing case, the maximum is `max(<length_dependent_factor>, max_time_num)`. The simplest option would be to introduce a possibility to allow configuring this accordingly. E.g. by something like `max_time_num_seq_len_divisor` which could be `int(1.0 / 0.7 * max_time)` like in the default case.

Reply:
But this would be the lower limit for the maximum number of masks. In the old implementation, it was possible to set `max_time_num` higher than this and cover the entire input. That is why I said minimum. I can add a `max_time_num_seq_len_divisor ` but I am not sure why I need the `max_time_num` parameter at all if the mask number is most of the time just dependent on the input length.

Reply:
> I am not sure why I need the max_time_num parameter at all if the mask number is most of the time just dependent on the input length.

That's a valid question. I think it would be good to have some statistics on the actual masking that is applied in training. I suggest you allow the implementation here to replicate the existing behaviour. Then you come up with a way to collect the mentioned statistics and then we can draw conclusions on how future experiments should be configured.

Reply:
Following up on this: Can you please also change the experiments introduced in this PR to a minimal baseline that is just reproducing the old behavior? This could serve as a sanity check and at the same time, we could collect said statistics. With the statistics available, you can then add a set of experiments varying some of the factors.

Reply:
I added an experiment that is very close to the old baseline, but I can not replicate the `increase_flag` behavior, since the scheduling can only change the sizes from one epoch to the next and not inside the first epoch.

Reply:
You could run both the current version and a version where you apply the behavior of the old `increase_flag` for the full first subepoch.